text/0000-sound-generic-drop.md

```diff
@@ -0,0 +1,1093 @@
+outlive the `Sneetch<'a>`" explicitly; Rust currently has no way of
+expressing the constraint that one lifetime be strictly greater than
+another lifetime or type (the form `'a:'b` only formally says that
+`'a` must live *at least* as long as `'b`).
+
+Thus, client code like that in `unwary_client` can inadvertantly set
+up scenarios where Sylvester's code may break, and Sylvester might be
+completely unaware of the vulnerability.
+
+## Type parameterization: the problem of trait bounds
+[The Zook example]: #type-parameterization-the-problem-of-trait-bounds
+
+One might think that all instances of this problem can
+be identified by the use of a lifetime-parametric `Drop` implementation,
+such as `impl<'a> Drop for Sneetch<'a> { ..> }`
+
+However, consider this trait and struct:
+
+```rust
+trait Button { fn push(&self); }
+struct Zook<B: Button> { button: B, }
+#[unsafe_destructor]
+impl<B: Button> Drop for Zook<B> {
+    fn drop(&mut self) { self.button.push(); }
+}
+```
+In this case, it is not obvious that there is anything wrong here.
+
+But if we continue the example:
+```rust
+struct Bomb { usable: bool }
+impl Drop for Bomb { fn drop(&mut self) { self.usable = false; } }
+impl Bomb { fn activate(&self) { assert!(self.usable) } }
+
+enum B<'a> { HarmlessButton, BigRedButton(&'a Bomb) }
+impl<'a> Button for B<'a> {
+    fn push(&self) {
+        if let B::BigRedButton(borrowed) = *self {
+            borrowed.activate();
+        }
+    }
+}
+
+fn main() {
+    let (mut zook, ticking);
+    zook = Zook { button: B::HarmlessButton };
+    ticking = Bomb { usable: true };
+    zook.button = B::BigRedButton(&ticking);
+}
+```
+Within the `zook` there is a hidden reference to borrowed data,
+`ticking`, that is assigned the same lifetime as `zook` but that
+will be dropped before `zook` is.
+
+(These examples may seem contrived; see [Appendix A] for a far less
+contrived example, that also illustrates how the use of borrowed data
+can lie hidden behind type parameters.)
+
+## The proposal
+
+This RFC is proposes to fix this scenario, by having the compiler
+ensure that types with destructors are only employed in contexts where
+either any borrowed data with lifetime `'a` within the type either
+strictly outlives the value of that type, or such borrowed data is
+provably not accessible from any `Drop` implementation via a reference
+of type `&'a`/`&'a mut`. This is the "Drop-Check" (aka `dropck`) rule.
+
+# Detailed design
+
+## The Drop-Check Rule
+[The Drop-Check Rule]: #the-drop-check-rule
+
+The Motivation section alluded to the compiler enforcing a new rule.
+Here is a more formal statement of that rule:
+
+Let `v` be some value (either temporary or named); if the type of `v` owns data of type `D`,
+where `D` has a lifetime- or type-parametric `Drop` implementation and
+either:
+
+  * (A.) the `Drop impl` for `D` instantiates `D` at `'a`
+         directly, i.e. `D<'a>`, or,
+
+  * (B.) the `Drop impl` for `D` has some type parameter with a
```

Comment:
Also, many types have an implicit `X: Drop` bound, e.g. `Vec` 

```rustimpl<T> Drop for Vec<T> {
    fn drop(&mut self) {
        // This is (and should always remain) a no-op if the fields are
        // zeroed (when moving out, because of #[unsafe_no_drop_flag]).
        if self.cap != 0 {
            unsafe {
                for x in self.iter() {
                    ptr::read(x); // reading the Ts, and running their destructors
                }
                dealloc(*self.ptr, self.cap)
            }
        }
    }
}
```